### PR TITLE
Add sycl event wait overload

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Add `dpctl.SyclEvent.wait()` overload for `dpjit` functions.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

Closes: #1138

### CODE REVIEW

File: numba_dpex/dpctl_iface/_intrinsic.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from numba.extending import intrinsic, overload_method
+
+import numba_dpex.dpctl_iface.libsyclinterface_bindings as sycl
+from numba_dpex.core import types as dpex_types
+
+
+@intrinsic
+def sycl_event_wait(typingctx, ty_event):
+    # check for accepted types
+    if not isinstance(ty_event, dpex_types.DpctlSyclEvent):
```
Comment (diptorupd) [inline numba_dpex/dpctl_iface/_intrinsic.py:17]: The check should be in the overload and a `TypingError` should be raised.

File: numba_dpex/dpctl_iface/_intrinsic.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        raise TypeError(f"Expected dpctl.SyclEvent, but got {ty_event}.")
+
+    result_type = types.void
+    sig = result_type(ty_event)
+
+    # defines the custom code generation
+    def codegen(context, builder, signature, args):
+        event_struct_proxy = cgutils.create_struct_proxy(ty_event)(
```
Comment (diptorupd) [inline numba_dpex/dpctl_iface/_intrinsic.py:25]: The change will not generate the correct code.

What you are doing is allocating a new empty struct for the `DpctlSyclEvent` type and then extracting the `event_ref` from that empty struct that will be a `Null` pointer. 

Following is the code your intrinsic generates:


```llvm
define i32 @"_ZN10numba_dpex11dpctl_iface10_intrinsic24ol_dpctl_sycl_event_wait12_3clocals_3e12_3clambda_3eB2v2B42c8tJTIcFKzyF2ILShI4CrgQElcaMVJhEEUSJJgA_3dE14DpctlSyclEvent"(i8** noalias nocapture %"retptr", {i8*, i32, i8*, i8*, i32}** noalias nocapture %"excinfo", i8* %"arg.event.0", i8* %"arg.event.1")
{
entry:
  %"inserted.parent" = insertvalue {i8*, i8*} undef, i8* %"arg.event.0", 0
  %"inserted.event_ref" = insertvalue {i8*, i8*} %"inserted.parent", i8* %"arg.event.1", 1
  %".6" = alloca {i8*, i8*} ; event_struct_proxy
  store {i8*, i8*} zeroinitializer, {i8*, i8*}* %".6"
B0:
  %"extracted.parent" = extractvalue {i8*, i8*} %"inserted.event_ref", 0
  %"extracted.event_ref" = extractvalue {i8*, i8*} %"inserted.event_ref", 1
  store {i8*, i8*} zeroinitializer, {i8*, i8*}* %".6"
  %".9" = load {i8*, i8*}, {i8*, i8*}* %".6" ; builder.load(event_struct_ptr)
  %".10" = extractvalue {i8*, i8*} %".9", 1 ; builder.extract_value
  %".11" = call i8* @"DPCTLEvent_Wait"(i8* %".10") ; sycl.dpctl_event_wait(builder, event_ref)
}
```

Instead you want to use `i8* %"arg.event.1"` which is the last argument to the `codegen` functions `args` argument.

```python
# defines the custom code generation
def codegen(context, builder, signature, args):
    sycl_event_dm = default_manager.lookup(ty_event)
    event_ref = builder.extract_value(
        args[0],
        sycl_event_dm.get_field_position("event_ref"),
    )
    sycl.dpctl_event_wait(builder, event_ref)
```

That way your generated code will be:

```llvm
define i32 @"_ZN10numba_dpex11dpctl_iface10_intrinsic24ol_dpctl_sycl_event_wait12_3clocals_3e12_3clambda_3eB2v2B42c8tJTIcFKzyF2ILShI4CrgQElcaMVJhEEUSJJgA_3dE14DpctlSyclEvent"(i8** noalias nocapture %"retptr", {i8*, i32, i8*, i8*, i32}** noalias nocapture %"excinfo", i8* %"arg.event.0", i8* %"arg.event.1")
{
entry:
  %"inserted.parent" = insertvalue {i8*, i8*} undef, i8* %"arg.event.0", 0
  %"inserted.event_ref" = insertvalue {i8*, i8*} %"inserted.parent", i8* %"arg.event.1", 1
B0:
  %"extracted.parent" = extractvalue {i8*, i8*} %"inserted.event_ref", 0
  %"extracted.event_ref" = extractvalue {i8*, i8*} %"inserted.event_ref", 1
  %".6" = extractvalue {i8*, i8*} %"inserted.event_ref", 1
  %".7" = call i8* @"DPCTLEvent_Wait"(i8* %".6")
}

```

Comment (ZzEeKkAa) [reply numba_dpex/dpctl_iface/_intrinsic.py]: Thank you!

**Review by diptorupd — APPROVED** [review]
Thank you! Looks good to me.

**Pull request merged**